File: openapi-clients/scripts/bump_versions.py
#!/usr/bin/env python3
import json
import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

TS_PACKAGE_JSON = ROOT / "typescript" / "package" / "package.json"
PYPROJECT_TOML = ROOT / "python" / "package" / "pyproject.toml"
RUST_CARGO_TOML = ROOT / "rust" / "package" / "Cargo.toml"


def bump_patch(version: str) -> str:
    match = re.match(r"^(\d+)\.(\d+)\.(\d+)$", version.strip())
    if not match:
        raise ValueError(f"Unsupported version format: {version}")
    major, minor, patch = map(int, match.groups())
    return f"{major}.{minor}.{patch + 1}"


def bump_package_json(path: Path) -> str:
    data = json.loads(path.read_text())
    current = data.get("version")
    if not current:
        raise ValueError(f"Missing version in {path}")
    data["version"] = bump_patch(current)
    path.write_text(json.dumps(data, indent=2) + "\n")
    return data["version"]


def bump_toml_version(path: Path) -> str:
    content = path.read_text()
    pattern = re.compile(r'^(version\s*=\s*")(\d+)\.(\d+)\.(\d+)(")', re.MULTILINE)
    match = pattern.search(content)
    if not match:
        raise ValueError(f"Missing version in {path}")
    current = ".".join(match.group(i) for i in range(2, 5))
    bumped = bump_patch(current)
    updated = pattern.sub(rf'\1{bumped}\5', content, count=1)
    path.write_text(updated)
    return bumped


def main() -> None:
    ts_version = bump_package_json(TS_PACKAGE_JSON)
    py_version = bump_toml_version(PYPROJECT_TOML)
    rust_version = bump_toml_version(RUST_CARGO_TOML)
    print(f"TypeScript version -> {ts_version}")
    print(f"Python version     -> {py_version}")
    print(f"Rust version       -> {rust_version}")


if __name__ == "__main__":
    main()

File: openapi-clients/bump_versions.ts
import fs from "fs";
import path from "path";
import toml, { JsonMap } from "@iarna/toml";

const ROOT = path.resolve(__dirname, "..");
const TS_PACKAGE_JSON = path.join(ROOT, "typescript", "package", "package.json");
const PYPROJECT_TOML = path.join(ROOT, "python", "package", "pyproject.toml");
const RUST_CARGO_TOML = path.join(ROOT, "rust", "package", "Cargo.toml");

type VersionTuple = [number, number, number];

function versionTuple(version: string): VersionTuple {
  const match = /^(\d+)\.(\d+)\.(\d+)$/.exec(version.trim());
  if (!match) {
    throw new Error(`Unsupported version format: ${version}`);
  }
  return [Number(match[1]), Number(match[2]), Number(match[3])];
}

function compareVersions(a: string, b: string): number {
  const aParts = versionTuple(a);
  const bParts = versionTuple(b);
  for (let i = 0; i < aParts.length; i += 1) {
    if (aParts[i] !== bParts[i]) {
      return aParts[i] - bParts[i];
    }
  }
  return 0;
}

function bumpPatch(version: string): string {
  const [major, minor, patch] = versionTuple(version);
  return `${major}.${minor}.${patch + 1}`;
}

function readPackageJsonVersion(filePath: string): string {
  const data = JSON.parse(fs.readFileSync(filePath, "utf8")) as { version?: string };
  if (!data.version) {
    throw new Error(`Missing version in ${filePath}`);
  }
  return data.version;
}

function writePackageJsonVersion(filePath: string, version: string): void {
  const data = JSON.parse(fs.readFileSync(filePath, "utf8")) as { version?: string };
  data.version = version;
  fs.writeFileSync(filePath, `${JSON.stringify(data, null, 2)}\n`, "utf8");
}

function readTomlVersion(filePath: string, table: string): string {
  const content = fs.readFileSync(filePath, "utf8");
  const data = toml.parse(content) as JsonMap;
  const tableData = data[table] as JsonMap | undefined;
  const version = tableData?.version;
  if (typeof version !== "string") {
    throw new Error(`Missing version in ${filePath}`);
  }
  return version;
}

function writeTomlVersion(filePath: string, table: string, version: string): void {
  const content = fs.readFileSync(filePath, "utf8");
  const data = toml.parse(content) as JsonMap;
  const tableData = (data[table] ?? {}) as JsonMap;
  tableData.version = version;
  data[table] = tableData;
  const updated = toml.stringify(data);
  fs.writeFileSync(filePath, updated.endsWith("\n") ? updated : `${updated}\n`, "utf8");
}

function main(): void {
  const tsCurrent = readPackageJsonVersion(TS_PACKAGE_JSON);
  const pyCurrent = readTomlVersion(PYPROJECT_TOML, "project");
  const rustCurrent = readTomlVersion(RUST_CARGO_TOML, "package");

  const versions = [tsCurrent, pyCurrent, rustCurrent];
  const maxCurrent = versions.reduce((max, current) =>
    compareVersions(current, max) > 0 ? current : max
  );
  const bumped = bumpPatch(maxCurrent);

  writePackageJsonVersion(TS_PACKAGE_JSON, bumped);
  writeTomlVersion(PYPROJECT_TOML, "project", bumped);
  writeTomlVersion(RUST_CARGO_TOML, "package", bumped);

  console.log(`TypeScript version -> ${bumped}`);
  console.log(`Python version     -> ${bumped}`);
  console.log(`Rust version       -> ${bumped}`);
}

main();
